lib/emulator.cpp

```diff
@@ -227,49 +225,43 @@ void emulator::emulator_t::check_and_handle_interrupts() {
 
     irvelog(1, "Checking for interrupts...");
 
-    Reg mstatus = this->m_CSR.implicit_read(Csr::Address::MSTATUS);
-    bool in_m_mode = this->m_CSR.get_privilege_mode() == PrivilegeMode::MACHINE_MODE;
-    bool in_s_mode = this->m_CSR.get_privilege_mode() == PrivilegeMode::SUPERVISOR_MODE;
-    
-    if (in_m_mode) {
-        if (!mstatus.bit(3)) {//mstatus.MIE is not set, so since we're in M-mode, this means interrupts are disabled
-            irvelog(0, "Interrupts are currently disabled in M-mode");
-            return;
-        }
-    } else if (in_s_mode) {
-        if (!mstatus.bit(1)) {//mstatus.SIE is not set, so since we're in S-mode, this means interrupts are disabled
-            irvelog(0, "Interrupts are currently disabled in S-mode");
-            return;
-        }
-    }
-    //If we reach this point, interrupts aren't globally disabled, so we need to check for them
+    auto interrupt_regs = this->m_CSR.fast_implicit_read_interrupt_regs();
 
-    //
-    //Check if any interrupts are "interrupting", and choose the one with the highest priority
-    //
+    Reg mstatus = interrupt_regs.mstatus;
+    bool in_m_mode = interrupt_regs.privilege_mode == PrivilegeMode::MACHINE_MODE;
+    bool in_s_mode = interrupt_regs.privilege_mode == PrivilegeMode::SUPERVISOR_MODE;
+
+    //NOTE if the interrupt is for a higher privilege, it does not matter if the global enable is unset for the current privilege
+    bool global_enable_for_current_privilege = (in_m_mode && (mstatus.bit(3) == 1)) || (in_s_mode && (mstatus.bit(1) == 1));
 
     //Get mip and sie. NOTE: We don't need to read sip and sie, since those are just shadows for S-mode code to use
```

Comment:
Yeah I'll remove it, this is some leftovers from what was previously there. :)

Reply:
I was more so referring to the line which was previously on 251:
```
//Get mip and sie. NOTE: We don't need to read sip and sie, since those are just shadows for S-mode code to use
```
Aren't we getting **mip** and **mie**?

Also, IMO the note would make more logical sense to be moved to the `InterruptRegs` struct definition since that's where we actually define which registers are needed for interrupts (especially if the lines below it are removed anyways)

Reply:
Ah good catch, I'll change that.

This comment was also present in this function from before I made these changes so that's why it wasn't in the struct definition.